File: platform/sdk-examples/typescript/dataset-workflow-example.ts
#!/usr/bin/env node
/**
 * Respan Dataset Workflow Example
 *
 * This example demonstrates the complete dataset workflow:
 * 1. List available logs
 * 2. Create a new dataset
 * 3. Add logs to the dataset
 * 4. List datasets
 * 5. Update dataset
 * 6. Run evaluation on dataset
 * 7. Get evaluation results
 *
 * Usage:
 *     npm run build && node dist/examples/dataset-workflow-example.js
 *     or
 *     npx tsx examples/dataset-workflow-example.ts
 *
 * Environment variables required:
 * - RESPAN_API_KEY
 * - RESPAN_BASE_URL (optional, defaults to http://localhost:8000)
 */

import { config } from 'dotenv';
import {
  DatasetAPI,
  EvaluatorAPI,
  type DatasetCreate,
  type DatasetUpdate,
  type LogManagementRequest,
} from '../src/index.js';

// Load environment variables with override
config({ override: true });

async function main() {
  // Setup
  const apiKey = process.env.RESPAN_API_KEY;
  const baseUrl = process.env.RESPAN_BASE_URL || 'http://localhost:8000';

  if (!apiKey) {
    console.log('❌ RESPAN_API_KEY not found in environment');
    console.log('   Please set your API key in .env file');
    return;
  }

  console.log('🚀 Respan Dataset Workflow Example');
  console.log('='.repeat(50));
  console.log(`🔗 Using API: ${baseUrl}`);
  console.log();

  // Initialize clients
  const datasetApi = new DatasetAPI({ apiKey, baseUrl });
  const evaluatorApi = new EvaluatorAPI({ apiKey, baseUrl });

  let createdDataset: any = null;

  try {
    // Step 1: List available logs for dataset creation
    console.log('📋 Step 1: Listing available logs...');
    const weekAgo = new Date();
    weekAgo.setDate(weekAgo.getDate() - 7);
    const now = new Date();

    // For demo purposes, we'll simulate having logs
    // In real usage, you'd get logs from the logs API
    const logsResponse = {
      results: [
        { id: 'log-1' },
        { id: 'log-2' },
        { id: 'log-3' }
      ]
    };

    console.log(`   ✅ Found ${logsResponse.results?.length || 0} logs`);
    let logIds: string[] = [];
    if (logsResponse.results && logsResponse.results.length > 0) {
      logIds = logsResponse.results.slice(0, 5).map((log: any) => log.id);
      console.log(`   📄 Sample log IDs: ${logIds.slice(0, 3).join(', ')}...`);
    } else {
      console.log('   ⚠️  No logs found in the specified time range');
    }
    console.log();

    // Step 2: Create a new dataset
    console.log('📝 Step 2: Creating a new dataset...');
    // Note: You can also use a simple dictionary instead of DatasetCreate!
    // See examples/flexible-input-example.ts for dictionary-based usage
    const datasetData: DatasetCreate = {
      name: `EXAMPLE_DATASET_${Math.floor(now.getTime() / 1000)}`,
      description: 'Example dataset created by SDK workflow',
      type: 'sampling',
      sampling: 10,
      start_time: weekAgo.toISOString(),
      end_time: now.toISOString()
    };

    createdDataset = await datasetApi.create(datasetData);
    console.log(`   ✅ Created dataset: ${createdDataset.name}`);
    console.log(`   🆔 Dataset ID: ${createdDataset.id}`);
    console.log(`   📊 Type: ${createdDataset.type}, Sampling: ${createdDataset.sampling || 'N/A'}`);
    console.log();

    // Step 3: Add logs to dataset (if we have logs)
    if (logIds.length > 0) {
      console.log('➕ Step 3: Adding logs to dataset...');
      const logRequest: LogManagementRequest = {
        start_time: weekAgo.toISOString().replace('T', ' ').slice(0, -5),
        end_time: now.toISOString().replace('T', ' ').slice(0, -5),
        filters: {
          id: {
            value: logIds,
            operator: 'in'
          }
        }
      };

      const addResult = await datasetApi.addLogsToDataset(createdDataset.id, logRequest);
      console.log(`   ✅ ${addResult.message || 'Logs added'}`);
      if (addResult.count !== undefined) {
        console.log(`   📊 Logs processed: ${addResult.count}`);
      }
      console.log();
    } else {
      console.log('⏭️  Step 3: Skipping log addition (no logs available)');
      console.log();
    }

    // Step 4: List all datasets
    console.log('📋 Step 4: Listing all datasets...');
    const datasets = await datasetApi.list({ pageSize: 5 });
    console.log(`   ✅ Found ${datasets.results.length} datasets`);
    for (let i = 0; i < Math.min(3, datasets.results.length); i++) {
      const ds = datasets.results[i];
      console.log(`   ${i + 1}. ${ds.name} (ID: ${ds.id})`);
    }
    console.log();

    // Step 5: Update the dataset
    console.log('✏️  Step 5: Updating dataset name...');
    const updateData: DatasetUpdate = { 
      name: `${datasetData.name}_UPDATED` 
    };
    const updatedDataset = await datasetApi.update(createdDataset.id, updateData);
    console.log(`   ✅ Updated name: ${updatedDataset.name}`);
    console.log();

    // Step 6: List available evaluators
    console.log('🔍 Step 6: Discovering available evaluators...');
    const evaluators = await evaluatorApi.list({ pageSize: 5 });
    console.log(`   ✅ Found ${evaluators.results.length} evaluators`);

    if (evaluators.results.length > 0) {
      for (let i = 0; i < Math.min(3, evaluators.results.length); i++) {
        const evaluator = evaluators.results[i];
        console.log(`   ${i + 1}. ${evaluator.name} (slug: ${evaluator.slug})`);
      }
      console.log();

      // Step 7: Run evaluation on dataset
      console.log('🎯 Step 7: Running evaluation on dataset...');
      const evaluatorSlug = evaluators.results[0].slug;
      console.log(`   🔧 Using evaluator: ${evaluatorSlug}`);

      try {
        const evalResult = await datasetApi.runDatasetEvaluation(
          createdDataset.id,
          [evaluatorSlug]
        );
        console.log(`   ✅ ${evalResult.message || 'Evaluation started'}`);
        console.log();

        // Step 8: Check evaluation results
        console.log('📊 Step 8: Checking evaluation results...');
        const reports = await datasetApi.listEvaluationReports(createdDataset.id);
        console.log(`   ✅ Found ${reports.results.length} evaluation reports`);

        if (reports.results.length > 0) {
          for (let i = 0; i < Math.min(2, reports.results.length); i++) {
            const report = reports.results[i];
            console.log(`   ${i + 1}. Report ID: ${report.id}`);
            if (report.status) {
              console.log(`      Status: ${report.status}`);
            }
          }
        }
        console.log();

      } catch (evalError) {
        console.log(`   ⚠️  Evaluation failed (may be expected if dataset has no logs): ${evalError}`);
        console.log();
      }
    } else {
      console.log('   ⚠️  No evaluators available');
      console.log();
    }

    // Step 9: Get dataset details
    console.log('📖 Step 9: Retrieving final dataset details...');
    const finalDataset = await datasetApi.get(createdDataset.id);
    console.log(`   ✅ Dataset: ${finalDataset.name}`);
    console.log(`   📅 Created: ${finalDataset.created_at || 'N/A'}`);
    console.log(`   📝 Description: ${finalDataset.description}`);
    console.log();

    console.log('🎉 Workflow completed successfully!');

  } catch (error) {
    console.log(`❌ Error in workflow: ${error}`);
    if (error instanceof Error) {
      console.error(error.stack);
    }

  } finally {
    // Cleanup
    if (createdDataset) {
      try {
        console.log('\n🗑️  Cleaning up...');
        await datasetApi.delete(createdDataset.id);
        console.log('   ✅ Test dataset deleted');
      } catch (cleanupError) {
        console.log(`   ⚠️  Could not delete dataset: ${cleanupError}`);
      }
    }
  }
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch(console.error);
}

File: platform/sdk-examples/python/dataset_workflow_example.py
#!/usr/bin/env python3
"""
Respan Dataset Workflow Example

This example demonstrates the complete dataset workflow:
1. List available logs
2. Create a new dataset
3. Add logs to the dataset
4. List datasets
5. Update dataset
6. Run evaluation on dataset
7. Get evaluation results

Usage:
    python examples/dataset_workflow_example.py

Environment variables required:
- RESPAN_API_KEY
- RESPAN_BASE_URL (optional, defaults to http://localhost:8000)
"""

import asyncio
import os
from datetime import datetime, timedelta
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

from respan.datasets.api import DatasetAPI
from respan.evaluators.api import EvaluatorAPI
from respan_sdk.respan_types.dataset_types import (
    DatasetCreate,
    DatasetUpdate,
    LogManagementRequest,
)


async def main():
    """Main example workflow"""
    
    # Setup
    api_key = os.getenv("RESPAN_API_KEY")
    base_url = os.getenv("RESPAN_BASE_URL", "http://localhost:8000")
    
    if not api_key:
        print("❌ RESPAN_API_KEY not found in environment")
        print("   Please set your API key in .env file")
        return
    
    print("🚀 Respan Dataset Workflow Example")
    print("=" * 50)
    print(f"🔗 Using API: {base_url}")
    print()
    
    # Initialize clients
    dataset_api = DatasetAPI(api_key=api_key, base_url=base_url)
    evaluator_api = EvaluatorAPI(api_key=api_key, base_url=base_url)
    
    created_dataset = None
    
    try:
        # Step 1: List available logs for dataset creation
        print("📋 Step 1: Listing available logs...")
        week_ago = datetime.utcnow() - timedelta(days=7)
        now = datetime.utcnow()
        
        logs_response = await dataset_api.client.get("logs/", params={
            "start_time": week_ago.strftime("%Y-%m-%d %H:%M:%S"),
            "end_time": now.strftime("%Y-%m-%d %H:%M:%S"),
            "page_size": 10
        })
        
        print(f"   ✅ Found {len(logs_response.get('results', []))} logs")
        if logs_response.get('results'):
            log_ids = [log['id'] for log in logs_response['results'][:5]]
            print(f"   📄 Sample log IDs: {log_ids[:3]}...")
        else:
            print("   ⚠️  No logs found in the specified time range")
            log_ids = []
        print()
        
        # Step 2: Create a new dataset
        print("📝 Step 2: Creating a new dataset...")
        # Note: You can also use a simple dictionary instead of DatasetCreate!
        # See examples/flexible_input_example.py for dictionary-based usage
        dataset_data = DatasetCreate(
            name=f"EXAMPLE_DATASET_{int(now.timestamp())}",
            description="Example dataset created by SDK workflow",
            type="sampling",
            sampling=10,
            start_time=week_ago.isoformat() + "Z",
            end_time=now.isoformat() + "Z"
        )
        
        created_dataset = await dataset_api.acreate(dataset_data)
        print(f"   ✅ Created dataset: {created_dataset.name}")
        print(f"   🆔 Dataset ID: {created_dataset.id}")
        print(f"   📊 Type: {created_dataset.type}, Sampling: {getattr(created_dataset, 'sampling', 'N/A')}")
        print()
        
        # Step 3: Add logs to dataset (if we have logs)
        if log_ids:
            print("➕ Step 3: Adding logs to dataset...")
            log_request = LogManagementRequest(
                start_time=week_ago.strftime("%Y-%m-%d %H:%M:%S"),
                end_time=now.strftime("%Y-%m-%d %H:%M:%S"),
                filters={
                    "id": {
                        "value": log_ids,
                        "operator": "in"
                    }
                }
            )
            
            add_result = await dataset_api.aadd_logs_to_dataset(created_dataset.id, log_request)
            print(f"   ✅ {add_result.get('message', 'Logs added')}")
            if 'count' in add_result:
                print(f"   📊 Logs processed: {add_result['count']}")
            print()
        else:
            print("⏭️  Step 3: Skipping log addition (no logs available)")
            print()
        
        # Step 4: List all datasets
        print("📋 Step 4: Listing all datasets...")
        datasets = await dataset_api.alist(page_size=5)
        print(f"   ✅ Found {len(datasets.results)} datasets")
        for i, ds in enumerate(datasets.results[:3], 1):
            print(f"   {i}. {ds.name} (ID: {ds.id})")
        print()
        
        # Step 5: Update the dataset
        print("✏️  Step 5: Updating dataset name...")
        update_data = DatasetUpdate(name=f"{dataset_data.name}_UPDATED")
        updated_dataset = await dataset_api.aupdate(created_dataset.id, update_data)
        print(f"   ✅ Updated name: {updated_dataset.name}")
        print()
        
        # Step 6: List available evaluators
        print("🔍 Step 6: Discovering available evaluators...")
        evaluators = await evaluator_api.alist(page_size=5)
        print(f"   ✅ Found {len(evaluators.results)} evaluators")
        
        if evaluators.results:
            for i, evaluator in enumerate(evaluators.results[:3], 1):
                print(f"   {i}. {evaluator.name} (slug: {evaluator.slug})")
            print()
            
            # Step 7: Run evaluation on dataset
            print("🎯 Step 7: Running evaluation on dataset...")
            evaluator_slug = evaluators.results[0].slug
            print(f"   🔧 Using evaluator: {evaluator_slug}")
            
            try:
                eval_result = await dataset_api.arun_dataset_evaluation(
                    created_dataset.id, 
                    [evaluator_slug]
                )
                print(f"   ✅ {eval_result.get('message', 'Evaluation started')}")
                print()
                
                # Step 8: Check evaluation results
                print("📊 Step 8: Checking evaluation results...")
                reports = await dataset_api.alist_evaluation_reports(created_dataset.id)
                print(f"   ✅ Found {len(reports.results)} evaluation reports")
                
                if reports.results:
                    for i, report in enumerate(reports.results[:2], 1):
                        print(f"   {i}. Report ID: {report.id}")
                        if hasattr(report, 'status'):
                            print(f"      Status: {report.status}")
                print()
                
            except Exception as eval_error:
                print(f"   ⚠️  Evaluation failed (may be expected if dataset has no logs): {eval_error}")
                print()
        else:
            print("   ⚠️  No evaluators available")
            print()
        
        # Step 9: Get dataset details
        print("📖 Step 9: Retrieving final dataset details...")
        final_dataset = await dataset_api.aget(created_dataset.id)
        print(f"   ✅ Dataset: {final_dataset.name}")
        print(f"   📅 Created: {getattr(final_dataset, 'created_at', 'N/A')}")
        print(f"   📝 Description: {final_dataset.description}")
        print()
        
        print("🎉 Workflow completed successfully!")
        
    except Exception as e:
        print(f"❌ Error in workflow: {e}")
        import traceback
        traceback.print_exc()
        
    finally:
        # Cleanup
        if created_dataset:
            try:
                print("\n🗑️  Cleaning up...")
                await dataset_api.adelete(created_dataset.id)
                print("   ✅ Test dataset deleted")
            except Exception as cleanup_error:
                print(f"   ⚠️  Could not delete dataset: {cleanup_error}")


if __name__ == "__main__":
    asyncio.run(main())